packages/mdc-textfield/foundation.js

```diff
@@ -142,6 +150,24 @@ class MDCTextFieldFoundation extends MDCFoundation {
     this.receivedUserInput_ = true;
   }
 
+  /**
+   * Updates the focus outline for outlined text fields.
+   */
+  updateOutline() {
```

Comment:
It bugs me that all of this logic has to be exposed at the top level, though I realize that it's probably impossible to isolate everything reasonably within the outline subcomponent (e.g. getting label size).

One thing I'm curious about is whether it'd be sufficient to get the width/height of the svg element rather than the root element, and if at least that part of this could be isolated within outline so you have less to compute and pass from here? I'd also suspect that `getIdleOutlineStyleValue` could be moved within the subcomponent too.

It's quite possible I'm not thinking this all the way through and don't know what you already tried, so let me know if this doesn't seem at all feasible.

Reply:
Good idea to move getting width/height into the outline subcomponent, I'll update that.
However `getIdleOutlineStyleValue` can't be moved to the subcomponent because the `idle-outline` element is outside of the subcomponent's root `outline` element, same with `labelWidth` and `isRtl`.